abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,919 @@
+(`map_type`).
+
+
+### `proxy_remove_map_value`
+
+* params:
+  - `i32 (proxy_map_type_t) map_type`
+  - `i32 (const char*) key_data`
+  - `i32 (size_t) key_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Remove key (`key_data`, `key_size`) from a given map (`map_type`).
+
+
+### `proxy_get_property`
+
+* params:
+  - `i32 (const char*) property_path_data`
+  - `i32 (size_t) property_path_size`
+  - `i32 (const char**) return_property_value_data`
+  - `i32 (size_t*) return_property_value_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Get property (`property_path_data`, `property_path_size`).
+
+
+### `proxy_set_property`
+
+* params:
+  - `i32 (const char*) property_path_data`
+  - `i32 (size_t) property_path_size`
+  - `i32 (const char*) property_value_data`
+  - `i32 (size_t) property_value_size`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Set property (`property_path_data`, `property_path_size`) to given value (`property_value_data`,
+`property_value_size`).
+
+
+## TCP/UDP/QUIC stream (L4) extensions
+
+### `proxy_resume_downstream`
+
+* params:
+  - none
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Resume processing of paused downstream.
+
+
+### `proxy_resume_upstream`
+
+* params:
+  - none
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Resume processing of paused upstream.
+
+
+## HTTP (L7) extensions
+
+### `proxy_resume_http_request`
+
+* params:
+  - none
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Resume processing of paused HTTP request.
+
+
+### `proxy_resume_http_response`
+
+* params:
+  - none
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Resume processing of paused HTTP response.
+
+
+### `proxy_send_http_response`
+
+* params:
+  - `i32 (uint32_t) response_code`
+  - `i32 (const char*) response_code_details_data`
+  - `i32 (size_t) response_code_details_size`
+  - `i32 (const char*) response_body_data`
+  - `i32 (size_t) response_body_size`
+  - `i32 (const char*) additional_headers_map_data`
+  - `i32 (size_t) additional_headers_size`
+  - `i32 (uint32_t) grpc_status`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Sends HTTP response without forwarding request to the upstream.
+
+
+## HTTP calls
+
+### `proxy_dispatch_http_call`
+
+* params:
+  - `i32 (const char*) upstream_name_data`
+  - `i32 (size_t) upstream_name_size`
+  - `i32 (const char*) headers_map_data`
+  - `i32 (size_t) headers_map_size`
+  - `i32 (const char*) body_data`
+  - `i32 (size_t) body_size`
+  - `i32 (const char*) trailers_map_data`
+  - `i32 (size_t) trailers_map_size`
+  - `i32 (uint32_t) timeout_milliseconds`
+  - `i32 (uint32_t*) return_callout_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Dispatch a HTTP call to upstream (`upstream_name_data`, `upstream_name_size`). Once the response is
+returned to the host, `proxy_on_http_call_response` will be called with a unique call identifier
```

Comment:
IIUC, in envoy there is no gaurantee that the http callbacks will not be called in-line with the http call. this is documented for failure case: 

https://github.com/envoyproxy/envoy/blob/95bbc547b8eea39cbab65a58ad49aafd957c14d6/include/envoy/http/async_client.h#L266-L268

but i see filters written defensively and assume that the success callback might be called inline as well. @mattklein123 might know for sure.

if the callbacks are called inline, the calling filter won't know the `callout_id` as it was not returned yet.

two things that come to mind:
- if only the failure callback can be called inline, then it is probably ok, as this is also indicated by the `call_result` return value
- if there is not such guarantee, then perhaps:
  - the calling filter should provide the callout_id ?
  - or perhaps the runtime can guarantee that `proxy_dispatch_http_call` will return before `proxy_on_http_call_response` is called, even if envoy  invokes the callbacks inline?

Reply:
IIRC it's not possible for the success callback to be called inline (I would need to check) but from an ABI perspective it probably makes sense to make this consistent per @yuval-k suggestions.